quil-rs/src/stats/execution_graph.rs

```diff
@@ -0,0 +1,200 @@
+use std::collections::HashMap;
+
+use crate::instruction::{Instruction, InstructionHandler, InstructionRole};
+use crate::quil::Quil;
+use petgraph::{graph::DiGraph, Direction};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Unsupported instruction: {}", .0.to_quil_or_debug())]
+    UnsupportedInstruction(Instruction),
+}
+
+/// ExecutionGraph is a logical execution/dependency graph of instructions. (Currently only 'gate' is supported.) It is a directed graph *from* the first instructions (the set of instructions that do not depend on prior instructions) *to* the last instructions (the set of instructions that are not prerequisites for any later instructions).
+pub struct ExecutionGraph {
+    graph: DiGraph<Instruction, ()>,
+}
+
+impl ExecutionGraph {
+    pub fn new(instructions: impl IntoIterator<Item = Instruction>) -> Result<Self, Error> {
+        let mut last_instruction_for_qubit = HashMap::new();
+        let mut graph = DiGraph::new();
+        let mut handler = InstructionHandler::default();
+
+        for instruction in instructions.into_iter() {
+            match handler.role_for_instruction(&instruction) {
+                InstructionRole::ClassicalCompute => {
+                    if let Instruction::Pragma(_) = instruction {
+                        return Err(Error::UnsupportedInstruction(instruction));
+                    }
+                } // Valid, mostly ignored
+                InstructionRole::ControlFlow => match &instruction {
+                    Instruction::Jump(_)
+                    | Instruction::JumpWhen(_)
+                    | Instruction::JumpUnless(_) => {
+                        return Err(Error::UnsupportedInstruction(instruction))
+                    }
+                    _ => {}
+                },
+                InstructionRole::ProgramComposition => {} // Valid, includes Gate, etc.,
+                InstructionRole::RFControl => {
+                    return Err(Error::UnsupportedInstruction(instruction))
+                }
+            }
+
+            let qubits: Vec<_> = instruction.get_qubits().into_iter().cloned().collect();
+
+            let node = graph.add_node(instruction);
+
+            for qubit in qubits {
+                if let Some(last_instruction) = last_instruction_for_qubit.insert(qubit, node) {
+                    graph.add_edge(last_instruction, node, ());
+                }
+            }
+        }
+
+        Ok(Self { graph })
+    }
+
+    // TODO: TEST THIS WELL
```

Comment:
This is implicitly tested via tests on public methods that use it. Can probably remove the TODO.